doc/source/ray-core/examples/rdt/grpo_contextual_bandits.py

```diff
@@ -0,0 +1,432 @@
+        return was_allowed
+
+    def allow_generation(self) -> bool:
+        was_allowed = self.generation_allowed.is_set()
+        self.generation_allowed.set()
+        return was_allowed
+
+    def is_generation_allowed(self) -> bool:
+        is_allowed = self.generation_allowed.is_set()
+        return is_allowed
+
+    def wait_for_generation(self, timeout: float | None = None) -> bool:
+        """Block until generation is allowed or timeout expires."""
+        return self.generation_allowed.wait(timeout)
+
+
+@ray.remote(num_gpus=1)
+class Generator:
+    """Generates actions using the current policy and sends them to scoring."""
+
+    def __init__(self, scorer, generation_signal) -> None:
+        self.model = MLP().to("cuda").eval()
+        self.scorer = scorer
+        self._generation_signal = generation_signal
+        self.policy_version = 1
+
+    @contextmanager
+    def _pause_generation(self):
+        """Temporarily block new generation requests while we swap weights."""
+        # SignalActor uses a threading.Event, so this round-trip keeps the pause scoped.
+        ray.get(self._generation_signal.disable_generation.remote())
+        try:
+            yield
+        finally:
+            # Always re-enable to avoid deadlocking the driver on failures.
+            ray.get(self._generation_signal.allow_generation.remote())
+
+    @ray.method(tensor_transport="nixl")
+    def generate(self, states: torch.Tensor):
+        ray.get(self._generation_signal.wait_for_generation.remote())
+        with torch.no_grad():
+            states_cuda = states.cuda()
+            logits = self.model(states_cuda)  # [batch_size, ACTION_DIM]
+
+            # GRPO requires sampling from the current policy (not just the greedy action)
+
+            # Create distribution for each state (batch_size distributions over ACTION_DIM actions each)
+            dist = Categorical(logits=logits)
+            # Sample ACTION_DIM actions from each state's distribution.
+            acts = dist.sample((ACTION_DIM,))  # [ACTION_DIM, batch_size]
+            # Vectorized log_prob over the sampled actions for each state
+            logps = dist.log_prob(acts)  # [ACTION_DIM, batch_size]
+            acts = acts.transpose(0, 1).contiguous()  # [batch_size, ACTION_DIM]
+            logps = logps.transpose(0, 1).contiguous()  # [batch_size, ACTION_DIM]
+
+        # Create trajectory slices and enqueue trajectory slices for Scorer
+        slice_batch = {
+            "policy_version": self.policy_version,
+            "state": states,
+            "actions": acts,
+            "old_logps": logps,
+        }
+        self.scorer.enqueue_trajectory_batch.remote(slice_batch)
+
+    def update_weights(self, cuda_weights, version: int) -> bool:
+        """Apply GPU-to-GPU policy weight updates while pausing sampling."""
+        # Disable sampling while the GPU object transfer and weight load are in-flight.
+        with self._pause_generation():
```

Comment:
I don't think we need this. The Generator is single-threaded and the single controller loop orders the tasks on the Generator, so it should never be in the middle of generation during a weight update anyway.